Add occ command to repair mtime

artonge opened the pull request:
`occ` command to repair files' `mtime`.

This is an early draft for another approach of fixing the files' `mtime`. It should be easier to update the `mtime` and to adapt it to different backends.

It's a modified version of the `Scan` command, so there is currently a lot of probably useless code.

This is to fix https://github.com/nextcloud-gmbh/server/issues/226#issuecomment-1054577115

artonge commented on apps/files/lib/Command/RepairMtime.php:
```diff
@@ -0,0 +1,311 @@
+
+		do {
+			$invalidFiles = $userFolder
+				// ->get($path)
```
This would be to handle the `--path` flag. How to get a `Folder` from a path?

artonge commented on apps/files/lib/Command/RepairMtime.php:
```diff
@@ -0,0 +1,311 @@
+					$filePath = $file->getPath();
+					$this->filesCounter++;
+					$storage = $file->getStorage();
+					$storageClass = get_class($storage);
```
I get `OCA\Files_Trashbin\Storage` with this line. What am I missing to have the real backend?

PVince81 replied:
you likely have the storage backend

you should use `->instanceOfStorage()` to check the type

artonge commented on apps/files/lib/Command/RepairMtime.php:
```diff
@@ -0,0 +1,311 @@
+				->search(
+					new SearchQuery(
+						new SearchComparison(ISearchComparison::COMPARE_LESS_THAN_EQUAL, 'mtime', 86400),
+						100,
```
Checks for invalid mtime 100 files at a time.

artonge commented on apps/files/lib/Command/RepairMtime.php:
```diff
@@ -0,0 +1,311 @@
+							break;
+						case AmazonS3::class:
+							$output->writeln("S3 storage detected for $filePath");
+							// $file->touch();
```
For S3 as primary backend, it is apparently possible to recover the `mtime` from the bucket metadata with a HEAD request.

PVince81 replied:
isn't AmazonS3::class the external storage ?
I believe the S3 primary storage is another class

mgallien commented on apps/files/lib/Command/RepairMtime.php:
```diff
@@ -0,0 +1,249 @@
+		do {
+			$invalidFiles = $userFolder
+				// ->get($path)
+				->search(
+					new SearchQuery(
+						new SearchComparison(ISearchComparison::COMPARE_LESS_THAN_EQUAL, 'mtime', 86400),
+						100,
+						$fileOffset,
+						[new SearchOrder(ISearchOrder::DIRECTION_DESCENDING, 'mtime')],
+						$user
+					)
+				);
```
not sure how you can this into account but unfortunately some files might also have an mtime in far future if they have been synced on Linux with a negative mtime stored in database and using desktop client latest bugfix releases
I have https://github.com/nextcloud/desktop/pull/4319 to fix this (due to an obsolete Qt API that is using unsigned int to store time and an overflow due to the negative value)

mgallien replied:
@artonge 
I have been working on a simple bash script for a one time recovery of sort for S3 primary storage
the new mtime for files is read from S3 `Last-Modified` header when GETing a file
might be an idea to add support for that

artonge commented:
@icewind1991 could you review the S3 part?

Notable changes:

- Added a `headObject` method in all implementation of `IObjectStore`
- Changed `AmazonS3::headObject` to a public method: https://github.com/nextcloud/server/pull/31386/files#diff-d26910d637d44440b2531cc06422315ff2c0553d6f14084c06b382dd6d10ac79R151
- Not sure what to return in `StorageObjectStore::headObject`: https://github.com/nextcloud/server/pull/31386/files#diff-ffe6a701709c4e9656d18bb0adf4cc63ec8b5c5fe02e970551fb6f7b713d01afR96

artonge committed:
Add occ command to repair mtime

Signed-off-by: Louis Chemineau <louis@chmn.me>
